Sparkle/SUUpdateValidator.m

```diff
@@ -7,79 +7,72 @@
-@synthesize dsaSignature = _dsaSignature;
+@synthesize prevalidatedSignature = _prevalidatedSignature;
+@synthesize signatures = _signatures;
+@synthesize downloadPrevalidationFailed = _downloadPrevalidationFailed;
 @synthesize downloadPath = _downloadPath;
 
-- (instancetype)initWithDownloadPath:(NSString *)downloadPath dsaSignature:(NSString *)dsaSignature host:(SUHost *)host performingPrevalidation:(BOOL)performingPrevalidation
+- (instancetype)initWithDownloadPath:(NSString *)downloadPath signatures:(SUSignatures *)signatures host:(SUHost *)host
 {
     self = [super init];
     if (self != nil) {
-        BOOL canValidate;
-        BOOL prevalidatedDsaSignature;
-        if (performingPrevalidation) {
-            NSString *publicDSAKey = host.publicDSAKey;
-
-            if (publicDSAKey == nil) {
-                prevalidatedDsaSignature = NO;
-                SULog(SULogLevelError, @"Failed to validate update before unarchiving because no DSA key was found");
-            } else if (dsaSignature == nil) {
-                prevalidatedDsaSignature = NO;
-                SULog(SULogLevelError, @"Failed to validate update before unarchiving because no DSA signature was found");
-            } else {
-                prevalidatedDsaSignature = [SUDSAVerifier validatePath:downloadPath withEncodedDSASignature:dsaSignature withPublicDSAKey:host.publicDSAKey];
-                if (!prevalidatedDsaSignature) {
-                    SULog(SULogLevelError, @"DSA signature validation before unarchiving failed for update %@", downloadPath);
-                }
-            }
-
-            canValidate = prevalidatedDsaSignature;
-        } else {
-            prevalidatedDsaSignature = NO;
-            canValidate = YES;
-        }
-
-        _canValidate = canValidate;
-        _prevalidatedDsaSignature = prevalidatedDsaSignature;
         _downloadPath = [downloadPath copy];
-        _dsaSignature = [dsaSignature copy];
+        _signatures = signatures;
         _host = host;
     }
     return self;
 }
 
+- (BOOL)validateDownloadPath {
+    SUPublicKeys *publicKeys = self.host.publicKeys;
+    SUSignatures *signatures = self.signatures;
+
+    if (publicKeys.dsaPubKey == nil && publicKeys.ed25519PubKey == nil) {
+        SULog(SULogLevelError, @"Failed to validate update before unarchiving because no (Ed)DSA public key was found in the old app");
+    } else {
+        if ([SUSignatureVerifier validatePath:self.downloadPath withSignatures:signatures withPublicKeys:publicKeys]) {
+            self.prevalidatedSignature = YES;
+            return YES;
+        }
+        SULog(SULogLevelError, @"(Ed)DSA signature validation before unarchiving failed for update %@", self.downloadPath);
+    }
+    self.downloadPrevalidationFailed = YES;
+    return NO;
+}
+
 - (BOOL)validateWithUpdateDirectory:(NSString *)updateDirectory
 {
-    assert(self.canValidate);
+    if (self.downloadPrevalidationFailed) {
+        return NO;
```

Comment:
I believe this property should just be removed; the caller makes sure this case doesn't happen. The `-validateDownloadPath` refactor is nice.

Reply:
That'd be a valid change, yeah. I suppose the design keeps you from getting into a bad state, though? We could compromise on an assert.

(In case it wasn't clear, I didn't write this code myself, just copied it over from 1.x.)

Reply:
Yeah I know. There used to be an assert ;).

Last I recall (not looking at it at the moment), it was confusing to to me to reason about this extra ivar state.

Reply:
The previous code had a precondition documented in the header file as well it seems.

Reply:
I suppose everything still ought to work fine even if it's not called, so sure, I'll take it out.